File: Leetcode/Problem list/188.cpp
#include <bits/stdc++.h>
using namespace std;
int n;
bool xd;
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        n=prices.size();
        vector<vector<vector<int>>> dp(2,vector<vector<int>>(k+1,vector<int>(2,xd=0)));
        for(int j=1;j<=k;j++){
            dp[0][j][1]=-prices[0];
        }
        for(int i=1;i<n;i++){
            xd=1-xd;
            for(int j=1;j<=k;j++){
                dp[xd][j][0]=max({dp[1-xd][j][1]+prices[i],dp[1-xd][j][0]});
                dp[xd][j][1]=max(dp[1-xd][j-1][0]-prices[i],dp[1-xd][j][1]);
            }
        }
        return dp[xd][k][0];
    }
};

File: Leetcode/Problem list/188.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        xd=0
        n=len(prices)
        dp=[[[0 for _ in range(2)] for _ in range(k+1)] for _ in range(2)]
        for j in range(1,k+1):
            dp[0][j][1]=-prices[0]
        for i in range(1,n):
            xd=1-xd
            for j in range(1,k+1):
                dp[xd][j][0]=max({dp[1-xd][j][1]+prices[i],dp[1-xd][j][0]})
                dp[xd][j][1]=max(dp[1-xd][j-1][0]-prices[i],dp[1-xd][j][1])
        return dp[xd][k][0]